### DISCUSSION TITLE
Exportable baked_in ValidatorFunc(s)

### DISCUSSION THREAD

**Discussion by gitusertd** [go-playground/validator#1141] [Ideas]
**Problem & Idea:**

- More often than not, we need to use custom validation function using **validator.StructLevel**
- But because all baked_in ValidationFunc are not exportable, such as hasValue, hasLengthOf, hasMinOf, isEq, etc. Because they don't start with uppercase letters, they cannot be imported and used.
- As a consequence, in our custom validation function, we cannot use those functions, having to do it ourselves and rewriting everything in order to validate our fields ...

**Solution:**
- Just rename all ValidationFunc(s), such as those from [baked_in.go](https://github.com/go-playground/validator/blob/master/baked_in.go) to Uppercase-first-character names. So that users can import those individual functions for private use cases.

**Comment by gitusertd** [comment]
It seems that the problem above can be solved with the validate.Var function

Example:

```
var validateObj *validator.Validate 
// validateObj must be initialized with  `RegisterTagNameFunc `and `RegisterStructValidation` later on, 
// and cached on the server, before it can be used for custom StructLevel validations.

func customValidateModel(vsl validator.StructLevel) {
    ...
    if err := validateObj.Var(model.Email, "required,email"); err != nil {
	 vsl.ReportError(model.Email, "email", "Email", "required,email", model.Email)
    }
    ...
}
```

**Comment by deankarn** [comment]
Yep thats the correct way.

As a follow-up note, it sounds very easy to expose the internal functions, however it also greatly increases the surface area of the code and then also must maintain backward compatibility, limiting changes and optimizations that can be done transparently and safely remaining internal only.
